### PR TITLE
feat: Add Human readable descriptions

### LINKED ISSUE #2011: [EPIC] Human-readable transactions
## What is the feature about

Ethereum transactions are complex and there’s no easy way to summarize what a single transaction does in one sentence. We propose to create an API to decode contract calls into a short and simple piece of text.

## ToDos

- [x] #2448
- [x] #2449

## Designs/sketches

TBD

## Links

https://www.notion.so/safe-global/Human-readable-tx-descriptions-8ede96b840584c96a8c58b546a59d921

### LINKED ISSUE #2448: Human-readable descriptions v1
## What is the feature about

This feature is part of the Epic #2011 and aims to add a first version of human-readable descriptions on the CGW and Web UI.

### CGW
https://github.com/safe-global/safe-client-gateway-nest/pull/631
- Create a structure to define messages for ABI calls
- Define messages for ERC20 and Safe transactions
- Implement parser and components for messages
- Return human descriptions as a simple string

### Web UI
- Add human descriptions to nonce replacement form, notifications

## Links

https://www.notion.so/safe-global/Human-readable-tx-descriptions-8ede96b840584c96a8c58b546a59d921

### PR DESCRIPTION
**Pull request by usame-algan**
## What it solves

First part of the EPIC #2011 

Resolves #2448 

Depends on https://github.com/safe-global/safe-client-gateway-nest/pull/631

## How this PR fixes it

- Displays human-readable descriptions for nonce replacement and notifications

## ToDos

- [x] Update gateway-typescript-sdk type for `TransactionInfo`
- [x] Check remaining `dispatch` functions if they need a `txDescription`

## How to test it

1. Queue a transaction (transfer, settings change, custom transaction)
2. Create a new transaction
3. Open the nonce replacement form
4. Observe human-readable descriptions
5. If there is no human-readable description observe the same message as before
6. Sign or execute a transaction
7. Observe a human-readable description inside notifications and the notification center
8. If there is no human-readable description observe a generic fallback saying `Transaction` as the Notification title

## Screenshots
Screenshot 2023-08-11 at 15 41 52
Screenshot 2023-08-11 at 17 22 29

Screenshot 2023-08-11 at 13 18 53

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/config/constants.ts [hunk 1]
```diff
@@ -3,9 +3,8 @@ import chains from './chains'
 export const IS_PRODUCTION = !!process.env.NEXT_PUBLIC_IS_PRODUCTION
 export const IS_DEV = process.env.NODE_ENV === 'development'
 
-export const GATEWAY_URL_PRODUCTION =
-  process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
-export const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+export const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'http://localhost:3030'
```
Comment (iamacook) [inline src/config/constants.ts:6]: Not review related: did you change the default gateway port? It opens on `3000` for me.

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,4 +0,0 @@
-{
```
Comment (iamacook) [inline .vscode/settings.json:1]: Perhaps we should add this to the `.gitignore`?

File: src/config/constants.ts (hunk 1 above)
Comment (usame-algan) [reply src/config/constants.ts:6]: Yes, I used a custom port 3030 with `APPLICATION_PORT=3030 yarn start:dev`

File: src/components/tx/SignOrExecuteForm/hooks.ts [hunk 3]
```diff
@@ -18,10 +18,11 @@ import type { OnboardAPI } from '@web3-onboard/core'
 import { getSafeTxGas, getRecommendedNonce } from '@/services/tx/tx-sender/recommendedNonce'
 import useAsync from '@/hooks/useAsync'
 import { useUpdateBatch } from '@/hooks/useDraftBatch'
+import { type TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
 
 type TxActions = {
-  addToBatch: (safeTx?: SafeTransaction, origin?: string) => Promise<string>
-  signTx: (safeTx?: SafeTransaction, txId?: string, origin?: string) => Promise<string>
+  addToBatch: (safeTx?: SafeTransaction, origin?: string, humanDescription?: string) => Promise<string>
+  signTx: (safeTx?: SafeTransaction, txId?: string, origin?: string, humanDescription?: string) => Promise<string>
```
Comment (katspaugh) [inline src/components/tx/SignOrExecuteForm/hooks.ts:25]: Passing `humanDescription` to these two methods seems oddly specific. I would suggest passing the whole `txDetails`.

File: src/hooks/useTxNotifications.ts [hunk 4]
```diff
@@ -79,9 +76,11 @@ const useTxNotifications = (): void => {
         const txId = 'txId' in detail ? detail.txId : undefined
         const txHash = 'txHash' in detail ? detail.txHash : undefined
         const groupKey = 'groupKey' in detail && detail.groupKey ? detail.groupKey : txId || ''
+        const humanDescription = 'humanDescription' in detail ? detail.humanDescription : ''
 
         dispatch(
           showNotification({
+            title: humanDescription,
```
Comment (katspaugh) [inline src/hooks/useTxNotifications.ts:83]: What will it look like w/o a title?

File: src/services/tx/tx-sender/dispatch.ts [hunk 5]
```diff
@@ -50,7 +50,10 @@ export const dispatchTxProposal = async ({
     proposedTx = await proposeTx(chainId, safeAddress, sender, safeTx, safeTxHash, origin)
   } catch (error) {
     if (txId) {
-      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, { txId, error: asError(error) })
+      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, {
+        txId,
+        error: asError(error),
+      })
```
Comment (katspaugh) [inline src/services/tx/tx-sender/dispatch.ts:56]: Unnecessary formatting
```suggestion
      txDispatch(TxEvent.SIGNATURE_PROPOSE_FAILED, { txId, error: asError(error) })
```

File: src/hooks/useTxNotifications.ts [hunk 6]
```diff
@@ -16,23 +16,20 @@ import useSafeAddress from './useSafeAddress'
...
+  [TxEvent.SIGNATURE_PROPOSE_FAILED]: 'Failed to send signature. Please try again.',
+  [TxEvent.EXECUTING]: 'Confirm the execution in your wallet.',
+  [TxEvent.PROCESSING]: 'Validating...',
+  [TxEvent.PROCESSING_MODULE]: 'Validating module interaction...',
+  [TxEvent.ONCHAIN_SIGNATURE_REQUESTED]: 'Confirm on-chain signature in your wallet.',
+  [TxEvent.ONCHAIN_SIGNATURE_SUCCESS]: 'On-chain signature request confirmed.',
+  [TxEvent.PROCESSED]: 'Successfully validated. Indexing...',
+  [TxEvent.REVERTED]: 'Reverted. Please check your gas settings.',
+  [TxEvent.SUCCESS]: 'Successfully executed.',
+  [TxEvent.FAILED]: 'Failed.',
```
Comment (iamacook) [inline src/hooks/useTxNotifications.ts:32]: Why are these changes necessary? If there's no `humanDescription`, we now lose some context.

Comment (katspaugh) [reply src/hooks/useTxNotifications.ts:32]: `humanDescription` will be always there unless the feature is disabled for technical reasons.

Comment (iamacook) [reply src/hooks/useTxNotifications.ts:32]: If the signature hash does not match on the gateway the `humanDescription` will be `null`. You can test this with spending limits atm:

image

Comment (usame-algan) [reply src/hooks/useTxNotifications.ts:32]: Good point, I think we should add a fallback for the title. Maybe we can include the nonce, something like `Transaction #113`

File: src/hooks/useTransactionType.ts [hunk 7]
```diff
@@ -80,13 +84,13 @@ export const getTransactionType = (tx: TransactionSummary, addressBook: AddressB
...
           icon: tx.safeAppInfo.logoUri,
-          text: tx.safeAppInfo.name,
+          text: tx.txInfo.humanDescription || tx.safeAppInfo.name,
         }
       }
 
       return {
         icon: toAddress?.logoUri || '/images/transactions/custom.svg',
-        text: addressBookName || toAddress?.name || 'Contract interaction',
+        text: tx.txInfo.humanDescription || addressBookName || toAddress?.name || 'Contract interaction',
```
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/hooks/useTransactionType.ts:93]: I am reverting this file since this is something we wan't to include in v2 as discussed.

File: src/hooks/useTxNotifications.ts (hunk 4 above)
Comment (usame-algan) [reply src/hooks/useTxNotifications.ts:83]: I went through each tx event that we show a notification for and made sure to add a humanDescription there, either from the CGW response or an appropriate fallback i.e. `Transaction #<nonce>` or `Module transaction`

File: src/components/tx/SignOrExecuteForm/hooks.ts (hunk 3 above)
Comment (usame-algan) [reply src/components/tx/SignOrExecuteForm/hooks.ts]: I've changed the signature to accept a transaction object instead.

File: src/components/tx/SignOrExecuteForm/hooks.ts [hunk 8]
```diff
@@ -83,50 +85,55 @@ export const useTxActions = (): TxActions => {
...
         // The signature will be added once the on-chain signature is indexed
         const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin)).txId
-        await dispatchOnChainSigning(safeTx, id, onboard, chainId)
+        await dispatchOnChainSigning(safeTx, id, onboard, chainId, humanDescription)
         return id
       }
 
       // Otherwise, sign off-chain
-      const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId)
+      const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId, humanDescription)
```
Comment (iamacook) [inline src/components/tx/SignOrExecuteForm/hooks.ts:106]: Should we also pass the whole transaction object here instead of `humanDescription`?

File: src/config/constants.ts [hunk 9]
```diff
@@ -3,9 +3,8 @@ import chains from './chains'
...
+export const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'http://localhost:3030'
```
Comment (iamacook) [inline src/config/constants.ts:7]: I'm unable to test on production with these fallbacks still set.

File: src/services/tx/tx-sender/dispatch.ts [hunk 10]
```diff
@@ -191,37 +200,39 @@ export const dispatchBatchExecution = async (
     const provider = createWeb3(wallet.provider)
     result = await multiSendContract.contract.connect(provider.getSigner()).multiSend(multiSendTxData, overrides)
 
-    txs.forEach(({ txId }) => {
-      txDispatch(TxEvent.EXECUTING, { txId, groupKey })
+    txs.forEach(({ txId, txInfo: { humanDescription } }) => {
```
Comment (iamacook) [inline src/services/tx/tx-sender/dispatch.ts:203]: Do we need a fallback for this dispatcher?

**Comment by francovenica** [comment]
This message is the same as the one in stg currently. in this case I tried to execute a batch tx (2 fully signed tx in queue, and using the Execute batch button), but these tx are incompatible so is expected for the batch to fail:

Idk if something more user friendly can be set, or if the information in the tx means something for the user.
image

**Comment by francovenica** [comment]
just "send"? in the description it shows that it should show what token you are sending, how much and to who.
In this case I'm testing in Goerli, sending GoerliETH
image

File: src/components/tx/SignOrExecuteForm/hooks.ts (hunk 8 above)
Comment (usame-algan) [reply src/components/tx/SignOrExecuteForm/hooks.ts:106]: We have all the info to build a humanDescription outside of the dispatch functions so I would suggest we keep all the logic there and only pass the string to keep those functions as pure as possible. I would move the fallbacks outside as well.

File: src/services/tx/tx-sender/dispatch.ts [hunk 11]
```diff
@@ -49,10 +49,11 @@ export const dispatchTxProposal = async ({
   try {
     proposedTx = await proposeTx(chainId, safeAddress, sender, safeTx, safeTxHash, origin)
   } catch (error) {
+    const humanDescription = `Transaction #${safeTx.data.nonce}`
```
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/services/tx/tx-sender/dispatch.ts]: I've moved this outside of this dispatch call so we have the human description + fallback in one place.

File: src/services/tx/tx-sender/dispatch.ts (hunk 10 above)
Comment (usame-algan) [reply src/services/tx/tx-sender/dispatch.ts]: I've changed it to a static description `Batch transaction`. Imo it is excessive to show a notification for every transaction but this is something we could also change later.

File: src/services/tx/tx-sender/dispatch.ts [hunk 12]
```diff
@@ -316,6 +334,7 @@ export const dispatchTxRelay = async (
   safe: SafeInfo,
   txId: string,
   gasLimit?: string | number,
+  humanDescription?: string,
```
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/services/tx/tx-sender/dispatch.ts:339]: I've added the same fallback as for the execution

File: src/store/txHistorySlice.ts [hunk 13]
```diff
@@ -28,7 +28,11 @@ export const txHistoryListener = (listenerMiddleware: typeof listenerMiddlewareI
         const txId = result.transaction.id
 
         if (pendingTxs[txId]) {
-          txDispatch(TxEvent.SUCCESS, { txId, groupKey: pendingTxs[txId].groupKey })
+          txDispatch(TxEvent.SUCCESS, {
+            txId,
+            groupKey: pendingTxs[txId].groupKey,
+            humanDescription: result.transaction.txInfo?.humanDescription,
```
[the remark this answers is not in the document]
Comment (usame-algan) [reply src/store/txHistorySlice.ts]: Thanks, I've added a fallback.

**Comment by usame-algan** [comment]
@francovenica thanks for testing! I've fixed the "Send" issue. This was a regression, nice find! For Batch Executions I now added a static description so it should always say "Batch transaction" as the title.

**Comment by iamacook** [comment]
> @iamacook @katspaugh what do you think about removing all the custom fallbacks and only having one fallback to "Transaction" inside `useTxNotifications`. We would lose some information like the nonce but I don't think its too critical since thats how it was before.

Having a SSOT makes the most sense.

**Comment by francovenica** [comment]
I tried the tx that Usame shared with me.
I was able to see most of them

I wasn't able to trigger "ONCHAIN_SIGNATURE_REQUESTED" and "ONCHAIN_SIGNATURE_SUCCESS". I'm gonna throw that, these probably were used back when every signature was on chain, even the ones that proposed a tx without execution.

The issue I reported about the nonce replacement message now show the tx name correctly

check the size of the nonce input tx display for mobile, looks fine as well:
image

Looks good to me

**Comment by usame-algan** [comment]
> Found an issue a last minute. a error message that show a "GS013" error

Nice find! There was an issue with setting the fallback title. It should be fixed now and display `Transaction` at the top of the notification.

**Comment by francovenica** [comment]
Yes, it has it now:
image

**Comment by francovenica** [comment]
Ok, looks good to me. I can pass this to done now.
